File: packages/python/luna/resources/za_tools/address.py
"""Luna SDK - South African Address Utilities for Python."""

from __future__ import annotations

import re
from .types import SAAddress, SAProvince


# Province postal code ranges
POSTAL_CODE_RANGES: dict[SAProvince, list[tuple[int, int]]] = {
    "EC": [(5000, 5999), (6000, 6499)],
    "FS": [(9300, 9999)],
    "GP": [(1, 299), (1000, 2199)],
    "KZN": [(3000, 4499)],
    "LP": [(500, 999)],
    "MP": [(1100, 1399), (2200, 2499)],
    "NC": [(8000, 8999)],
    "NW": [(2500, 2999), (300, 499)],
    "WC": [(6500, 8099)],
}

# SA cities to province mapping
SA_CITIES: dict[str, SAProvince] = {
    "johannesburg": "GP",
    "pretoria": "GP",
    "cape town": "WC",
    "durban": "KZN",
    "port elizabeth": "EC",
    "gqeberha": "EC",
    "bloemfontein": "FS",
    "polokwane": "LP",
    "nelspruit": "MP",
    "mbombela": "MP",
    "kimberley": "NC",
    "rustenburg": "NW",
    "soweto": "GP",
    "sandton": "GP",
    "centurion": "GP",
}

# Province names
PROVINCE_NAMES: dict[SAProvince, str] = {
    "EC": "Eastern Cape",
    "FS": "Free State",
    "GP": "Gauteng",
    "KZN": "KwaZulu-Natal",
    "LP": "Limpopo",
    "MP": "Mpumalanga",
    "NC": "Northern Cape",
    "NW": "North West",
    "WC": "Western Cape",
}


class AddressUtils:
    """South African address utilities."""

    def validate(self, address: SAAddress) -> dict:
        """Validate a South African address."""
        errors: list[str] = []
        warnings: list[str] = []

        # Validate postal code
        if address.postal_code:
            if not re.match(r"^\d{4}$", address.postal_code):
                errors.append("Postal code must be 4 digits")
            else:
                code = int(address.postal_code)
                detected_province = self.get_province_from_postal_code(address.postal_code)
                if address.province and detected_province and address.province != detected_province:
                    warnings.append(
                        f"Postal code {address.postal_code} belongs to {PROVINCE_NAMES.get(detected_province, detected_province)}, "
                        f"but address shows {PROVINCE_NAMES.get(address.province, address.province)}"
                    )

        # Validate province
        if address.province and address.province not in PROVINCE_NAMES:
            errors.append(f"Invalid province code: {address.province}")

        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "warnings": warnings,
        }

    def format(self, address: SAAddress, multiline: bool = True) -> str:
        """Format address for display."""
        parts: list[str] = []

        if address.street:
            parts.append(address.street)
        if address.suburb:
            parts.append(address.suburb)
        if address.city:
            parts.append(address.city)
        if address.province:
            province_name = PROVINCE_NAMES.get(address.province, address.province)
            parts.append(province_name)
        if address.postal_code:
            parts.append(address.postal_code)
        parts.append(address.country)

        separator = "\n" if multiline else ", "
        return separator.join(parts)

    def get_province_from_postal_code(self, postal_code: str) -> SAProvince | None:
        """Detect province from postal code."""
        if not re.match(r"^\d{4}$", postal_code):
            return None

        code = int(postal_code)

        for province, ranges in POSTAL_CODE_RANGES.items():
            for range_start, range_end in ranges:
                if range_start <= code <= range_end:
                    return province

        return None

    def get_province_name(self, code: SAProvince) -> str:
        """Get full province name from code."""
        return PROVINCE_NAMES.get(code, code)

    def lookup_postal_code(self, postal_code: str) -> dict | None:
        """Look up postal code information."""
        if not re.match(r"^\d{4}$", postal_code):
            return None

        province = self.get_province_from_postal_code(postal_code)
        if not province:
            return None

        return {
            "postal_code": postal_code,
            "province": province,
            "province_name": PROVINCE_NAMES.get(province, province),
        }

    def parse(self, address_string: str) -> SAAddress:
        """Parse a free-form address string."""
        result = SAAddress(country="ZA")

        # Try to extract postal code
        postal_match = re.search(r"\b(\d{4})\b", address_string)
        if postal_match and postal_match.group(1):
            result.postal_code = postal_match.group(1)
            result.province = self.get_province_from_postal_code(postal_match.group(1))

        # Try to extract city
        for city, province in SA_CITIES.items():
            if city in address_string.lower():
                result.city = city.title()
                result.province = result.province or province
                break

        return result

    def title_case(self, text: str) -> str:
        """Convert text to title case."""
        return " ".join(word.capitalize() for word in text.split())

File: packages/go/luna/resources/zatools/address.go
// Package zatools provides South African business tool integrations.
package zatools

import (
	"regexp"
	"strings"
)

// AddressUtils provides South African address utilities.
type AddressUtils struct{}

// NewAddressUtils creates a new address utils instance.
func NewAddressUtils() *AddressUtils {
	return &AddressUtils{}
}

// Postal code ranges by province
var postalCodeRanges = map[SAProvince][][2]int{
	ProvinceEC:  {{5000, 5999}, {6000, 6499}},
	ProvinceFS:  {{9300, 9999}},
	ProvinceGP:  {{1, 299}, {1000, 2199}},
	ProvinceKZN: {{3000, 4499}},
	ProvinceLP:  {{500, 999}},
	ProvinceMP:  {{1100, 1399}, {2200, 2499}},
	ProvinceNC:  {{8000, 8999}},
	ProvinceNW:  {{2500, 2999}, {300, 499}},
	ProvinceWC:  {{6500, 8099}},
}

// SA cities to province mapping
var saCities = map[string]SAProvince{
	"johannesburg":   ProvinceGP,
	"pretoria":       ProvinceGP,
	"cape town":      ProvinceWC,
	"durban":         ProvinceKZN,
	"port elizabeth": ProvinceEC,
	"gqeberha":       ProvinceEC,
	"bloemfontein":   ProvinceFS,
	"polokwane":      ProvinceLP,
	"nelspruit":      ProvinceMP,
	"mbombela":       ProvinceMP,
	"kimberley":      ProvinceNC,
	"rustenburg":     ProvinceNW,
	"soweto":         ProvinceGP,
	"sandton":        ProvinceGP,
	"centurion":      ProvinceGP,
}

// Province names
var provinceNames = map[SAProvince]string{
	ProvinceEC:  "Eastern Cape",
	ProvinceFS:  "Free State",
	ProvinceGP:  "Gauteng",
	ProvinceKZN: "KwaZulu-Natal",
	ProvinceLP:  "Limpopo",
	ProvinceMP:  "Mpumalanga",
	ProvinceNC:  "Northern Cape",
	ProvinceNW:  "North West",
	ProvinceWC:  "Western Cape",
}

// Validate validates a South African address.
func (a *AddressUtils) Validate(address SAAddress) map[string]interface{} {
	var errors []string
	var warnings []string

	// Validate postal code
	if address.PostalCode != "" {
		if !regexp.MustCompile(`^\d{4}$`).MatchString(address.PostalCode) {
			errors = append(errors, "Postal code must be 4 digits")
		} else {
			detectedProvince := a.GetProvinceFromPostalCode(address.PostalCode)
			if address.Province != "" && detectedProvince != "" && address.Province != detectedProvince {
				warnings = append(warnings, "Postal code province mismatch")
			}
		}
	}

	// Validate province
	if address.Province != "" {
		if _, ok := provinceNames[address.Province]; !ok {
			errors = append(errors, "Invalid province code")
		}
	}

	return map[string]interface{}{
		"valid":    len(errors) == 0,
		"errors":   errors,
		"warnings": warnings,
	}
}

// Format formats address for display.
func (a *AddressUtils) Format(address SAAddress, multiline bool) string {
	var parts []string

	if address.Street != "" {
		parts = append(parts, address.Street)
	}
	if address.Suburb != "" {
		parts = append(parts, address.Suburb)
	}
	if address.City != "" {
		parts = append(parts, address.City)
	}
	if address.Province != "" {
		if name, ok := provinceNames[address.Province]; ok {
			parts = append(parts, name)
		} else {
			parts = append(parts, string(address.Province))
		}
	}
	if address.PostalCode != "" {
		parts = append(parts, address.PostalCode)
	}
	parts = append(parts, address.Country)

	separator := ", "
	if multiline {
		separator = "\n"
	}
	return strings.Join(parts, separator)
}

// GetProvinceFromPostalCode detects province from postal code.
func (a *AddressUtils) GetProvinceFromPostalCode(postalCode string) SAProvince {
	if !regexp.MustCompile(`^\d{4}$`).MatchString(postalCode) {
		return ""
	}

	var code int
	for _, c := range postalCode {
		code = code*10 + int(c-'0')
	}

	for province, ranges := range postalCodeRanges {
		for _, r := range ranges {
			if code >= r[0] && code <= r[1] {
				return province
			}
		}
	}

	return ""
}

// GetProvinceName gets full province name from code.
func (a *AddressUtils) GetProvinceName(code SAProvince) string {
	if name, ok := provinceNames[code]; ok {
		return name
	}
	return string(code)
}

// LookupPostalCode looks up postal code information.
func (a *AddressUtils) LookupPostalCode(postalCode string) map[string]interface{} {
	if !regexp.MustCompile(`^\d{4}$`).MatchString(postalCode) {
		return nil
	}

	province := a.GetProvinceFromPostalCode(postalCode)
	if province == "" {
		return nil
	}

	return map[string]interface{}{
		"postal_code":   postalCode,
		"province":      province,
		"province_name": provinceNames[province],
	}
}

// Parse parses a free-form address string.
func (a *AddressUtils) Parse(addressString string) SAAddress {
	result := SAAddress{Country: "ZA"}

	// Try to extract postal code
	postalMatch := regexp.MustCompile(`\b(\d{4})\b`).FindStringSubmatch(addressString)
	if len(postalMatch) >= 2 {
		result.PostalCode = postalMatch[1]
		result.Province = a.GetProvinceFromPostalCode(postalMatch[1])
	}

	// Try to extract city
	lower := strings.ToLower(addressString)
	for city, province := range saCities {
		if strings.Contains(lower, city) {
			result.City = strings.Title(city)
			if result.Province == "" {
				result.Province = province
			}
			break
		}
	}

	return result
}
